cli_tools_e2e_test/gce_windows_upgrade/test_suites/windows_upgrade/windows_upgrade_test_suite.go

```diff
@@ -0,0 +1,448 @@
+					// rollback case
+					if containsAdditionalLicense {
+						utils.Failure(testCase, logger, "Additional license found after rollback.")
+					}
+				} else {
+					// cleanup case
+					if !containsAdditionalLicense {
+						utils.Failure(testCase, logger, "Additional license not found.")
+					}
+				}
+			}
+		}
+
+		hasBootDisk = true
+	}
+	if !hasBootDisk {
+		utils.Failure(testCase, logger, "Boot disk not found.")
+		return
+	}
+	if len(instance.Disks) != dataDiskCount + 1 {
+		utils.Failure(testCase, logger, fmt.Sprintf("Count of disks not match: expect %v, actual %v.", dataDiskCount+1, len(instance.Disks)))
+	}
+
+	if expectSuccess {
+		// for success case, verify OS version by startup script
+		err = instance.RestartWithScript("$ver=[System.Environment]::OSVersion.Version\n" +
+			"Write-Host \"windows_upgrade_verify_version=$($ver.Major).$($ver.Minor)\"")
+		if err != nil {
+			testCase.WriteFailure("Error starting instance `%v` with script: `%v`", instanceName, err)
+			return
+		}
+		expectedOutput := "windows_upgrade_verify_version=6.3"
+		logger.Printf("[%v] Waiting for `%v` in instance serial console.", instanceName,
+			expectedOutput)
+		if err := instance.WaitForSerialOutput(
+			expectedOutput, 1, 15*time.Second, 15*time.Minute); err != nil {
+			testCase.WriteFailure("Error during validation: %v", err)
+		}
+	} else {
+		// for failed case, verify rollback
+		if autoRollback {
+			// original boot disk name == instance name by default
+			originalOSDisk, err := instance.Client.GetDisk(testProjectConfig.TestProjectID, testProjectConfig.TestZone, instanceName)
+			if err != nil {
+				utils.Failure(testCase, logger, "Failed to get original OS disk.")
+			}
+			if len(originalOSDisk.Users) == 0 {
+				utils.Failure(testCase, logger, "Original OS disk didn't get rollback.")
+			}
+		}
+	}
+
+	// for all cases, verify cleanup: install media, startup script & backup
+	for _, d := range instance.Disks {
+		if strings.HasSuffix(d.Source, "global/images/family/windows-install-media") {
+			utils.Failure(testCase, logger, "Install media is not cleaned up.")
+		}
+	}
+	var windowsStartupScriptURL string
+	var windowsStartupScriptURLBackup string
+	for _, i := range instance.Metadata.Items {
+		if i.Key == "windows-startup-script-url" && i.Value != nil {
+			windowsStartupScriptURL = *i.Value
+		} else if i.Key == "windows-startup-script-url-backup" && i.Value != nil {
+			windowsStartupScriptURLBackup = *i.Value
+		}
+	}
+	if expectedScriptURL != windowsStartupScriptURL {
+		utils.Failure(testCase, logger, fmt.Sprintf("Unexpected startup script URL: %v", windowsStartupScriptURL))
+	}
+	if windowsStartupScriptURLBackup != "" {
+		utils.Failure(testCase, logger, fmt.Sprintf("Unexpected startup script URL backup: %v", windowsStartupScriptURLBackup))
+	}
+}
+
+func containsSubString(strs []string, s string) bool {
+	for _, str := range strs {
+		if strings.Contains(str, s) {
+			return true
+		}
+	}
+	return false
+}
```

Comment:
Seems like a common func that could be reused in both test and prod code. Consider extracting to a util file.